File: tarjan_scc.cpp
#include <iostream>
#include <vector>
#include <functional>
std::vector<std::vector<int>> tarjanSCC(std::vector<std::vector<int>> &adj) {
	int n = adj.size(), clock = 0;
	std::vector<int> low(n), dis(n); // 'dis'covery time and 'low'est node reachable
	std::vector<int> has(n), stk; // 'has' checks if 'stk' has a node
	std::vector<std::vector<int>> sccs; // holds a list of sccs with their nodes
	std::function<void (int)> dfs = [&](int u) {
		dis[u] = low[u] = ++clock;
		stk.push_back(u);
		has[u] = true;
		for (int v : adj[u]) {
			if (!dis[v])
				dfs(v);
			if (has[v]) // On the stack means found lower
				low[u] = std::min(low[u], low[v]);
		}
		if (low[u] == dis[u]) { // No path to low nodes, so it's the cutoff for an SCC
			sccs.push_back({});
			while (true) {
				int v = stk.back();
				stk.pop_back();
				has[v] = false;
				sccs.back().push_back(v);
				if (u == v)
					break;
			}
		}
	};
	for (int u = 0; u < adj.size(); ++u) // assumes zero indexed
		if (!dis[u])
			dfs(u);
	return sccs;
}

File: tarjan_scc.py
def tarjanSCC(adj: [[int]]):
	n, clock = len(adj), iter(range(1, len(adj)+1))
	dis, low = [0]*n, [0]*n # 'dis'covery time and 'low'est node reachable
	has, stk = [0]*n, [] # 'has' checks if a node is on the DFS 'stk'
	sccs: [[int]] = [] # holds a list of sccs with their nodes
	def dfs(u: int):
		dis[u] = low[u] = next(clock)
		stk.append(u)
		has[u] = True
		for v in adj[u]:
			if not dis[v]:
				dfs(v)
			if has[v]: # on the stack means found lower
				low[u] = min(low[u], low[v])
		if low[u] == dis[u]: # did not find lower
			sccs.append([])
			while True:
				v = stk.pop()
				has[v] = False;
				sccs[-1].append(v)
				if v == u:
					break
	for u in range(n):
		if not dis[u]:
			dfs(u)
	return sccs
